src/html.rs

```diff
@@ -0,0 +1,92 @@
+use std::collections::{BTreeMap, BTreeSet};
+
+use once_cell::sync::Lazy;
+use scraper::{ElementRef, Html, Selector};
+use serde::Deserialize;
+
+#[derive(Debug, Deserialize)]
+struct Config<'a> {
+    #[serde(borrow)]
+    bad_sections: BTreeMap<&'a str, BTreeSet<&'a str>>,
+}
+
+static CONFIG: Lazy<Config<'static>> = Lazy::new(|| {
+    serde_json::from_str(include_str!(concat!(
+        env!("CARGO_MANIFEST_DIR"),
+        "/article_processing_config.json"
+    )))
+    .expect("\"article_processing_config.json\" is either invalid json or the wrong structure")
+});
+
+static HEADERS: Lazy<Selector> =
+    Lazy::new(|| Selector::parse("h1, h2, h3, h4, h5, h6, h7").unwrap());
+
+pub fn simplify(html: &str, lang: &str) -> String {
+    let mut document = Html::parse_document(html);
+
+    let mut to_remove = Vec::new();
+
+    // remove sections
+
+    if let Some(bad_sections) = CONFIG.bad_sections.get(lang) {
+        for header in document.select(&HEADERS) {
+            // TODO: should this join all text nodes?
+            let Some(title) = header.text().next() else {
+                continue
+            };
+
+            if bad_sections.contains(&title) {
+                to_remove.push(header.id());
+                let header_level = header.value().name();
+                // strip trailing nodes
+                for sibling in header.next_siblings() {
+                    if let Some(element) = sibling.value().as_element() {
+                        if element.name() == header_level {
+                            // TODO: should this check for a higher level?
+                            break;
+                        }
+                    }
+                    to_remove.push(sibling.id());
+                }
+            }
+        }
+
+        for id in to_remove.drain(..) {
+            if let Some(mut node) = document.tree.get_mut(id) {
+                node.detach();
+            }
+        }
+    } else {
+        warn!("No bad sections configured for lang {lang:?}");
+    }
+
+    // remove elements with no text that isn't whitespace
+
+    for element in document
+        .root_element()
+        .descendants()
+        .filter_map(ElementRef::wrap)
+    {
+        if element.text().all(|t| t.trim().is_empty()) {
+            to_remove.push(element.id());
+        }
+    }
+
+    for id in to_remove.drain(..) {
+        if let Some(mut node) = document.tree.get_mut(id) {
+            node.detach();
+        }
+    }
```

Comment:
Can copy-paste be avoided?

Reply:
I'll be improving and refactoring this in the next PR, if this bit survives I'll move it to a function.